drivers/flash/soc_flash_nrf.h

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __SOC_FLASH_NRF_H__
+#define __SOC_FLASH_NRF_H__
+
+#include <kernel.h>
+
+#define FLASH_OP_DONE    (0) /* 0 for compliance with the driver API. */
+#define FLASH_OP_ONGOING (-1)
+
+struct flash_context {
+	uint32_t data_addr;  /* Address of data to write. */
+	uint32_t flash_addr; /* Address of flash to write or erase. */
+	uint32_t len;        /* Size off data to write or erase [B]. */
+#ifndef CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE
+	uint8_t  enable_time_limit; /* set execution limited to the execution
+				     * window.
+				     */
+	void     *synch_context;    /* timing context used by given
+				     * synchronization backend. It is used to
+				     * transfer timing parameters and store
+				     * data required by the backend.
+				     * Structure of contex depends on backend
+				     * implementation.
+				     */
+#endif /* !CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE */
+#if defined(CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE)
+	uint32_t flash_addr_next;
+#endif /* CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE */
+}; /*< Context type for f. @ref write_op @ref erase_op */
+
+#ifndef CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE
+/**
+ * @defgroup nrf_flash_synch synch backend API
+ *
+ * API declared bellow contains prototypes of function which shall be
+ * implemented by the synchronization backend.
+ * @{
+ */
+
+/**
+ * Calback which executes the flash operation.
+ *
+ * @param context pointer to flash_context structure.
+ * @retval @ref FLASH_OP_DONE once operation was done, @ref FLASH_OP_ONGOING if
+ *         opperation needs more time for execution.
+ */
+typedef int (*flash_op_handler_t) (void *context);
+
+struct flash_op_desc {
+	flash_op_handler_t handler;
+	struct flash_context *context; /* [in,out] */
+};
+
+/**
+ * Synchronization backend driver initialization procedure.
+ *
+ * This will be run within flash driver initialization
+ */
+int nrfsb_synch_backend_init(void);
+
+/**
+ * Give synchronization context for write operation.
+ *
+ * @retval Synchronization context for the write operation.
+ * Structure of contex depends on backend implementation.
+ */
+void *nrfsb_get_write_synch_context(void);
```

Comment:
I have an idea to remove this function and `nrfsb_get_erase_synch_context()` from API. Instead of preparing write or erase operation, the `soc_flash_nrf` module could pass expected duration of the next flash operation within `struct flash_op_desc` to the synchronization backend. The backend would add its overhead and recalculate timings units according to its requirements.

I understand that it could be less optimal, because synchronization backend would need to calculate operations duration in software for each flash operation. However it looks that API, call flow (number of branch instructions), modules structure would be simplified, what could result in similar performance. But for sure would result in easier maintenance of `soc_flash_nrf` and sync backends.

Any thoughts?

Reply:
I think this is not additional maintenance effort as by the build time configuration timing values (in scientific units) are converted using macros to beck-end implementation own values using macros provide by the ticker infrastructure. What I was thinking introducing per backed defined context was that it is possible that each of them might have original timing variables each other, even different number of orthogonal variables.

@cvinayak What your opinion.

Reply:
Let me rephrase to describe what I meant by increased maintenance.

Synchronization backend does not care if `soc_flash_nrf` performs flash write operation, flash erase, partial erase, partial write, bulk write, bulk erase, or any other operation that I can imagine. It only needs to know duration of the next expected operation. In the proposed interface, the duration is passed by calling either `nrfsb_get_write_synch_context()` or `nrfsb_get_erase_synch_context()`. Now it is responsibility of the backend to deduct expected flash operation timings based on last called function. Each backend would need to contain code that converts and stores expected operation duration based on a call to one or another function.

If the map of flash operation durations would be in `soc_flash_nrf`, it would be in single place. But if we would keep such maps in each backend, similar conversion is copied in multiple modules. Maintaining map in single module is easier than in multiple copies.

One example of increased maintenance would be porting `soc_flash_nrf` driver to a new device with changed flash operations durations. If operations durations were placed only in `soc_flash_nrf.c` and passed as an argument to backends, only 1 file would have to be modified while porting. But if we keep flash operation durations in sync backends, each backend would have to be updated, what is more work and greater possibility of mistake.

Another advantage of keeping duration in `soc_flash_nrf` in single file would be easy extension of the driver. Right now the driver supports only two operations in single product "write" and "erase". When a new use case comes to add "partial erase" operation to the same product that supports typical "erase" (given type of erase could be selected run-time based on number of simultaneous connection), extending `soc_flash_nrf` to add third operation would require changes only in `soc_flash_nrf`. But if we keep current interface, adding new flash operation would require change in `soc_flash_nrf`, alongside the interface and all backend implementations.

Reply:
@de-nordic and your opinion?

Reply:
I am getting towards the proposal.

Reply:
I agree with the proposal from @hubertmis - This will require less maintenance, and gives a more logical abstraction of scheduling back-end vs. flash timings.

Reply:
I am reworking the driver accordingly.